evm/src/cpu/kernel/asm/memory/syscalls.asm

```diff
@@ -213,3 +213,67 @@ global sys_returndatacopy:
 returndatacopy_empty:
     %stack (kexit_info, dest_offset, offset, size) -> (kexit_info)
     EXIT_KERNEL
+
+// Same as %wcopy but with special handling in case of overlapping ranges.
+global sys_mcopy:
+    // stack: kexit_info, dest_offset, offset, size
+    PUSH @GAS_VERYLOW
+    // stack: Gverylow, kexit_info, dest_offset, offset, size
+    DUP5 %num_bytes_to_num_words %mul_const(@GAS_COPY) ADD %charge_gas
+
+    %stack (kexit_info, dest_offset, offset, size) -> (dest_offset, size, kexit_info, dest_offset, offset, size)
+    %add_or_fault
+    // stack: expanded_num_bytes, kexit_info, dest_offset, offset, size, kexit_info
+    DUP1 %ensure_reasonable_offset
+    %update_mem_bytes
+
+    // stack:  kexit_info, dest_offset, offset, size
+    DUP4
+    // stack:  size, kexit_info, dest_offset, offset, size
+    ISZERO %jumpi(returndatacopy_empty) // If size is empty, just pop the stack and exit the kernel
```

Comment:
Very nit: can't we do the check before verifying that the offset is reasonable?

Reply:
Yes! I had originally copied the `returndatacopy` code which has it also there, but it can be moved before